Feature/recreate enumerable methods

AymenBida opened the pull request:
### Recreate enumerable methods

In this project, we tried to recreate some of the enumerable methods of Ruby such as (#map, #each, #select ...) using #yield and proc#call in the process

We can't correct Rubocop offenses because it's asking to use #each instead of #for and we are not allowed to do so in the requirements

AymenBida committed:
Add initial files and create #my_each method

Commit:
Create #my_each_with_index method

Leandro-Barretoo committed:
test commit

AymenBida committed:
Create #my_select method

AymenBida committed:
Create #my_all? method

AymenBida committed:
Create #my_any? and #my_none? methods

Leandro-Barretoo committed:
Add #my_count method

Leandro-Barretoo committed:
Create #my_map method

AymenBida committed:
Create #my_inject and #multiply_els methods

AymenBida committed:
Change #my_map method to be able to take a proc as argument

AymenBida committed:
Correct some rubocop offenses

AymenBida committed:
Edit the readme file

arslanbisharat commented on my_scripts.rb:
```diff
@@ -0,0 +1,94 @@
+    end
+  end
+
+  def my_each_with_index(var)
+    if block_given?
+      for i in 0...var.length
+        yield(var[i], i)
+      end
+      var
+    else 'Error: no block given'
+    end
+  end
```
**1-** my_each_with_index should returns an Enumerator if no block is given:
`p [1, 2, 3].my_each_with_index` returns `#<Enumerator>`
<img width="457" alt="2" src="https://user-images.githubusercontent.com/58504655/103807613-78ec9400-5078-11eb-9679-4420ee943788.PNG">

**2-**

AymenBida committed:
Correct the way the methods are called

AymenBida committed:
fixed Rubocop offenses

AymenBida commented:
### Enumerable

In this instance, we : 

- Corrected the way methods are called (`object.method` instead of `method(object)`)
- Fixed the Rubocop issues

ElijahTobs reviewed (changes requested):
# Status: Changes Required :recycle: 

## Hi @AymenBida and @Leandro-Barretoo :wave: 

You've done well so far :+1:  and I commend you for that :clap: 
However, there are still some changes to be made before this milestone can be approved.

## REQUIRED

Some of your custom enumerable methods did not failed some test cases. See them below:
* [x] my_all
* [x] my_any
* [x] my_none
* [x] my_count
* [x] my_inject
* [x] multiply_els
------------------------------------------------------------------------------------------------
**my_all**
When no block or argument is given my_all? should return true when all the elements in the enum and truthy values and return false if at least one element in the collection is a falsy value.
example [1, 2, 3].my_all? should return true and [nil, 1, 2, 3].my_all? should return false.

my_all? should be able to receive an optional argument. If the argument is a class, then it should return true if all members of the given enumerable are members of that class, otherwise it should return false. If the argument is a regex, then it should return true if all members of the enumerable matches the regex, otherwise it should return false. If the argument is any other pattern besides a class or regex, a similar logic holds.
Example

[1, 2, 3].my_all?(Integer) should return true
[1, 2, "a"].my_all?(Integer) should return false
['dog', 'door'].my_all?(/d/) should return true
['dog', 'door', 'ant'].my_all?(/d/) should return false
[2, 2, 2].my_all?(2) should return true
[1, 2, 3].my_all?(2) should return false

**my_any**
When no block or argument is given to my_any? it should return true if there is at least one element in the given enumerable that is a truthy, otherwise, it should return false.
example [1, 2, false].my_any? should return true and [nil, false nil].my_any? should return false.

my_any? should be able to receive an optional argument. If the argument is a class, then it should return true if there is at least one member of the given enumerable that is a member of that class, otherwise it should return false. If the argument is a regex, then it should return true if there is at least one member of the enumerable that matches the regex, otherwise it should return false. If the argument is any other pattern besides a class or regex, a similar logic holds.
Example

[1, 2, "a"].my_any?(Integer) should return true
["a",  "b",  "a"].my_any?(Integer) should return false
['dog', 'door', 'ant'].my_any?(/d/) should return true
['hey', 'book', 'ant'].my_any?(/d/) should return false
[2, 2, 3].my_any?(2) should return true
[1, 3, 3].my_any?(2) should return false

**my_none**
When no block or argument is given to my_none?, it should return true if there is no element in the given enumerable that is a truthy, otherwise, it should return false.
example [nil, nil, false].my_none? should return true and [1, 2, 3].my_none? should return false.

my_none? should be able to receive an optional argument. If the argument is a class, then it should return true if there is no member of the given enumerable that is a member of that class, otherwise it should return false. If the argument is a regex, then it should return true if there is no member of the enumerable that matches the regex, otherwise it should return false. If the argument is any other pattern besides a class or regex, a similar logic holds.
Example

["a", "b", "a"].my_none?(Integer) should return true
["1",  "2",  "a"].my_none?(Integer) should return false
['rot', 'not', 'ant'].my_none?(/d/) should return true
['door', 'dog', 'ant'].my_none?(/d/) should return false
[1, 3, 3].my_none?(2) should return true
[1, 2, 3].my_none?(2) should return false

**my_count**
It is supposed to count the number of items in enum that are equal to item if an argument is given

**my_inject**
Please read this article to understand how the inject method behaves and compare the behaviour with your own my_inject method.

**multiply_els**
It is supposed to accept an array as an argument and multiplies all the elements of the array together using #my_inject

### Please, make the requested changes and submit for another review.
## Happy Coding :computer: 

_Feel free to reach out to me on slack_ **`@Elijah Ayandokun`**
  **Elijah O. Ayandokun**
[Github](https://github.com/ElijahTobs) | [LinkedIn](https://www.linkedin.com/in/ayandokunelijah/) | [Twitter](https://twitter.com/ElijahTobs)

Leandro-Barretoo committed:
Fix my_all? method

AymenBida committed:
Correct #my_any? and #my_none? methods

AymenBida committed:
Correct #my_count method

AymenBida committed:
test commit

Leandro-Barretoo committed:
Fix #my_inject method and #multiply_els method

Leandro-Barretoo committed:
Fix #multiply_els method

AymenBida committed:
Fix rubocop issues

AymenBida commented:
### Enumerable methods

In this instance, we fixed the problems with :

- my_all
- my_any
- my_none
- my_count
- my_inject
- multiply_els

Georjane reviewed (changes requested):
## Status: Changes Required 🛑 🔧 
### Hi team @AymenBida and @Leandro-Barretoo ✋🏽 
Nice work on the enumerables. Your methods are now passing most of the test cases except the 2 following:

- [x] enumerables `#my_inject` raises a "LocalJumpError" when no block or argument is given
- [x] enumerables` #multiply_els` accepts an array as an argument and multiplies all the elements of the array together using #my_inject. Remember that the ` #multiply_els` is not an enumerable method, so you should remove it and put it outside the Enumerable modules.
Please kindly implement these changes and submit for another review.
Feel free to reach out to me on Slack if you have any questions, **_Witah Georjane._** 🤝 

Good luck on your project and Happy Coding 💻

AymenBida committed:
Fix #my_inject method so it raises a local jump error when no block or argument is given

AymenBida commented:
### Enumerable methods

In this instance, we did the following : 

- Change `#my_inject` so it now raises a local jump error when no block or argument is given
- Take `#multiply_els` out from the `#Enumerable` module

Georjane reviewed (approved):
## Status: PR Approved 💯 🥇 
### Hi Team ✋🏽 
Congratulations on successfully completing the enumerable methods. All test cases are now passing. It's a thumbs up for me. 👍🏽 You can go ahead and merge this pull request.

Good luck on your next project and Happy Coding 💻
